### PR TITLE
docs(eslint-plugin): clearly document extension rules

### LINKED ISSUE #1446: Add a mechanism for marking a rule as an extension rule
We have a number of rules which extend eslint functionality.

Currently, the only place that we list these extension rules is [here](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/tools/generate-configs.ts#L24-L42).
This is easy to forget about, and it is hard to expose to users.

We should do the following things:
- Similar to how we added the `requiresTypeChecking?: boolean` flag to the rule `meta`, we should add a new flag, say `baseRuleExtension?: boolean`.
- We should generate documentation based on this (maybe a separate table in the eslint-plugin readme, to clearly list them?).
- We should add a section to the [FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) about checking the extension rule list when you encounter a bug in a base rule.

### PR DESCRIPTION
**Pull request by bradzacher**
Fixes #1446

- Adds a flag to the rule meta.
- Moves the extension rules to a new table in the readme.
- Adds a section to the FAQ.
- Update doc checker accordingly.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-magic-numbers.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export default util.createRule<Options, MessageIds>({
       description: 'Disallow magic numbers',
       category: 'Best Practices',
       recommended: false,
+      extendsBaseRule: true,
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/no-magic-numbers.ts]: small question about this, do we want to support string here?

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/no-magic-numbers.ts]: that's a good point.
We already have this case with `no-return-await` (implemented as `return-await`)

Comment (armano2) [reply packages/eslint-plugin/src/rules/no-magic-numbers.ts]: maybe we should go one step further and replace it with string[] as naming-convention rule replaces a lot of eslint rules.

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/no-magic-numbers.ts]: `naming-convention` is a weird case... It's not an extension rule, it's a brand new rule which overlaps with some base rules.

I don't think we want to list this "overlap" in the docs anywhere.

----

As a side note, I was actually considering submitting `naming-convention` as a rule for eslint core.
The only typescript-specific thing in there is the types option, which is easy enough to trim out.

File: packages/eslint-plugin/tests/configs.test.ts [hunk 2]
```diff
@@ -14,10 +27,27 @@ function filterRules(values: Record<string, string>): [string, string][] {
...
+): void {
+  it('has the base rules overriden by the appropriate extension rules', () => {
+    const ruleNames = new Set(Object.keys(unfilteredConfigRules));
+    EXTENSION_RULES.forEach(([ruleName, extRuleName]) => {
+      if (ruleNames.has(ruleName)) {
+        // this looks a little weird, but it provides the cleanest test output style
+        expect(unfilteredConfigRules).toMatchObject({
+          ...unfilteredConfigRules,
+          [extRuleName]: 'off',
+        });
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/configs.test.ts]: image

**Pull request merged**